presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import io.prestosql.sql.planner.assertions.BasePlanTest;
+import org.testng.annotations.Test;
+
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.filter;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.output;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestUnwrapCastInComparison
+        extends BasePlanTest
+{
+    @Test
+    public void testEquals()
+    {
+        // representable
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '1') t(a) WHERE a = DOUBLE '1'",
+                anyTree(
+                        filter("A = SMALLINT '1'",
```

Comment:
Use different value in WHERE (filter) than in VALUES. Now the two `1`s are different, yet undistinguishable.